GPII-3125 Init GCP organization

amatas opened the pull request:
This code covers [GPII-3125](https://issues.gpii.net/browse/GPII-3125), [GPII-3127](https://issues.gpii.net/browse/GPII-3127), [GPII-3126](https://issues.gpii.net/browse/GPII-3126) and [GPII-3128](https://issues.gpii.net/browse/GPII-3128).

More info about how to apply this code is in the [README.md](https://github.com/gpii-ops/gpii-infra/blob/35b4ba9f9657b0c25e23c0f0d488d386453c7dd6/common/README.md)

This pr includes the following changes:

- Move the rakefiles to a common part shared by all the project types
- Create the DNS structure for a complete organization, in AWS and GCP
- The organization details are parametrized.
- Make idempotent the creation of the initial resources used by each environment
- Move some rake tasks to the Docker container
- `rake sh` gives a shell inside the container without making changes.

This is the plan to merge this PR:

1. Merge the PR :)
1. Let the CI test all until `stg`. This should create the initial resources for test1.gpii.net and create/destroy a complete dev testing cluster.
1. Manually import in the `prd` state file the projects already created in the organization in order to avoid naming conflicts and/or increase the projects quota.
1. Import the DNS zones that match the [DNS pattern](https://github.com/gpii-ops/gpii-infra/blob/d60a61a5ee8d03cd11fb205cfceedbba817f5945/shared/rakefiles/vars.rb#L28-L36), delete others.
1. Promote to `prd` to recreate all the initial resources, DNS zones left, ns registries, SA of each project, activate API services and update the TF state of the common part.
1. Each individual project should refresh the TF state simply running `rake`, the CI will do automatically.
1. Re-run the complete pipeline in order to check that the idempotency works as expected.

stepanstipl commented:
I'm wondering if ths is not overly complicated way to achieve what we want. I don't think it's a problem of this PR as such, but rather of the whole structure. I know some of this is enforced given the used tooing, but that would be imho another reason to consider simpler approach.

My main concerns being:
- **Many dependencies**, currently these are required (not limited to):
  - Ruby 
  - Rake - using Rakefiles
  - Terraform
  - Terragrunt
  - Custom fork of Exekube
  - Docker-compose
  - Docker (of course :))
  - Python (called from Tf)
  - gcloud cli
 
- **Complitaced and confusing dir structure** - `common/live/prd/infra/development/` for example, each of these containing significant amount of duplicated code (abusing DRY)

I think the complexity of this will backfire in future. In general I'm fan of Occam's razor principle & keep it simple.

As alternative I would propose using pure Terraform with a Makefile, all packaged in Docker container, gcloud would still have to be there. I think the code for our purpose could then simply look something like:

```tf
...
module "dev-tyler" {
  source  = "modules/gcp-project"

  name    = "dev-tyler"
  owner   = "tyler@raisingthefloor.org"
}

module "dev-stepan" {
  source  = "module/gcp-project"

  name    = "dev-stepan"
  owner   = "stepan@raisingthefloor.org"
}
...
```

This would make this piece of code more maintainable with less dependencies.

We can also create the DNS zone per project here together with corresponding DNS records. For exmple the subdomains for above code would be `.dev-tyler.gcp.gpii.net` and `.dev-stepan.gcp.gpii.net` and then the code for actual K8s infrastructure could create any records underneath.

(Probably, at least initially, I would also suggest leaving out the AWS DNS zone mgmt., and simply create one subdomain that would be "manually" delegated to GCP, as I assume we will want to get rid of AWS in future.)

amatas commented on common/modules/gcp-project/main.tf:
```diff
@@ -0,0 +1,89 @@
+# compute the project_id to get the dns zone in JSON format
+# I wasn't able to replicate this function using HCL code
+data "external" "calculate_dns_zone" {
+  program = ["python", "-c", "print(\"{\\\"zone\\\": \\\"\" + \".\".join(\"${google_project.project.project_id}\".split(\"-\")[::-1]) + \".net.\") + \"\\\"}\""]
```
```
$ python -c 'print("{\"zone\": \"" + ".".join("gpii-gcp-dev-alfredo".split("-")[::-1]) + ".net." + "\"}")'
{"zone": "alfredo.dev.gcp.gpii.net."}
```
it takes the name of the project `gpii-gcp-dev-alfredo` and returns the DNS zone in json format: `{"zone": "alfredo.dev.gcp.gpii.net."}`

The command split the project name by the hyphens and rejoin them in the other way around ([::-1]), adds a .net at the end and wrap all in {"zone": "dnsname"}.

I don't know how to do this in bash, I just wanted to reduce the parameters of the module. Perhaps using something like [this terraform function](https://github.com/hashicorp/terraform/issues/9404) but it's not implemented yet.
I used python because Exekube containers [have it installed](https://github.com/exekube/exekube/blob/master/dockerfiles/google/Dockerfile#L34)

Perhaps I can expand the comment about this resource in the .tf in order to understand what it does.

natarajaya replied:
@amatas This works:
```
variable "project_name" {
  default = "gpii-gcp-dev-alfredo"
}

output "project_zone" {
  value = "{\"zone\": \"${replace(var.project_name, "/([a-z]+)-([a-z]+)-([a-z]+)-([a-z]+)/", "$4.$3.$2.$1")}.\"}"
}
```

```
Apply complete! Resources: 4 added, 0 changed, 0 destroyed.

Outputs:

project_zone = {"zone": "alfredo.dev.gcp.gpii."}
```

amatas replied:
The possible input values are:
```
gpii-gcp-dev
gpii-gcp-dev-alfredo
```
this interpolation doesn't work with `gpii-gcp-dev` because it the length of the items is fixed.
```
variable "project_name" {
  default = "gpii-gcp-dev"
}

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

project_zone = {"zone": "gpii-gcp-dev."}
```

natarajaya replied:
@amatas This works for both variants:
```
variable "project_name" {
  default = "gpii-gcp-dev"
}

output "project_zone" {
  value = "{\"zone\": \"${
    replace(
      replace(
        var.project_name,
        "/([a-z]+)-([a-z]+)-([a-z]+)-?([a-z]+)?/",
        "$4.$3.$2.$1."),
      "/^\\./",
      "")
    }\"}"
}
```
Not sure if it is more readable than your solution, but at least it does not have any python in it :)

mrtyler replied:
I think Sergey's version is about as readable as the Python one-liner. (The part I found trickiest was removing the leading `.` when `$4` is undefined. :)) I would write a comment explaining the general problem (and maybe the leading dot thing) and adopt this solution.

natarajaya commented on common/docker-compose.yaml:
```diff
@@ -35,6 +35,9 @@ services:
       # The GCP KMS keyring name to use
       TF_VAR_keyring_name: keyring
 
+      TF_VAR_organization_id: ${ORGANIZATION_ID}
```
I can go further with this and say – can we maybe use the same docker compose everywhere when `exekube` is needed? :)

mrtyler replied:
The original motivation for having separate (and different) `docker-compose.yaml` was because the `common/` side needs `~/.aws` mounted in while the `gcp/` side does not. (I think there were 1-2 other differences as well, but I don't think I see them now. Perhaps the compose.env abstraction allows these differences to be hidden?)

If that mount is the only remaining difference between common/d-c.yaml and gcp/d-c.yaml, perhaps it is worth unifying the two files. `~/.aws` can be left (or forced to be) empty on the gcp/ side.

Alfredo, are we missing anything from the last time we discussed this?

natarajaya commented on common/live/prd/Rakefile:
```diff
@@ -0,0 +1,7 @@
+import "../../../gcp/rakefiles/entrypoint.rake"
```
+1 for making rakefiles common as well.

natarajaya commented on common/live/prd/infra/dev-natarajaya/terraform.tfvars:
```diff
@@ -0,0 +1,19 @@
+# ↓ Module metadata
```
I think this should be in `common/live/dev/natarajaya`.

natarajaya replied:
Nevermind, I replied with more details in https://github.com/gpii-ops/gpii-infra/pull/60#discussion_r205665333

natarajaya commented:
@amatas I know it took me so long to review this! My overall impression is good, I replied on some comments in discussions started by Tyler. I did not go into Ruby code yet.

Main of my points here:
* Shared rakefiles / docker-compose configs
* Live directory structure (https://github.com/gpii-ops/gpii-infra/pull/60#discussion_r205665333)
* Use Terraform native functions instead of Python (https://github.com/gpii-ops/gpii-infra/pull/60#discussion_r205667215)
* How to re-use dns zone created in dev project (https://github.com/gpii-ops/gpii-infra/pull/60/files#r205671169)
 
Another question – how are we going to test this if we can not delete our existing projects and let this code recreate everything from scratch? :)

amatas commented on common/live/prd/infra/dev-alfredo/terraform.tfvars:
```diff
@@ -0,0 +1,19 @@
+# ↓ Module metadata
```
Replied in https://github.com/gpii-ops/gpii-infra/pull/60#discussion_r205732233

amatas commented on common/modules/gcp-project/main.tf:
```diff
@@ -0,0 +1,89 @@
+  ]
+}
+ 
+resource "google_dns_managed_zone" "project" {
```
My first attempt would be to import the resource before running the `up` command in order to be "ignored" in the apply step. This could be the less disruptive change in the infrastructure. 

Other approaches could be either removing the resource from the gke-network module, or change the zone creation at this PR.

But I can't either test or evaluate the impact all of the above before deploying the project first.

natarajaya replied:
Ok, I did some testing. Looks like the easiest solution that does not require any changes to Terraform code – do not set dns records and zones here:
https://github.com/gpii-ops/gpii-infra/blob/master/gcp/rakefiles/vars.rb#L37

natarajaya commented on common/live/prd/infra/development/terraform.tfvars:
```diff
@@ -0,0 +1,19 @@
+# ↓ Module metadata
```
@amatas Check this out:
https://github.com/natarajaya/exekube-test
I created test exekube project with proposed `live` structure and it works great! No issues with dependencies.
cc @mrtyler

mrtyler replied:
I think the innovation that makes Sergey's version work is moving the code that creates zone resources out of `development/terraform.tfvars` (i.e. the directory itself) and into a module `development/root/terraform.tfvars`. I like that part.

However, I'm not sure I like the name `root`. Next to entries like `tyler` and `natarajaya`, it makes me think of the root user, not DNS. How about keeping the dedicated directories, but calling them `zone` instead?

I like changing the name of top-level `prd` to something else to reflect the unusual/singleton-y nature of the common/ side of gpii-infra. I was going to say that I don't like `gpii-net` since there are (or will be) more than DNS resources managed by this project, but I can't think of anything better. The project name looks reasonable to me as well: `gpii-common-gpii-net`. Maybe someday we will buy `gpii-test.net` where we can stage changes to production DNS. 

Any other ideas on names for this env?

```
common/
  live/
    gpii-net/
      zone/
      dev/
        zone/
        alfredo/
        natarajaya/
        tyler/
      stg/
        zone/
      prd/
        zone/
```

mrtyler commented:
@stepanstipl:

> DNS - do we need different DNS structure for the dev projects?

One motivation for creating a dedicated `dev` zone was indeed to be able to give it different permissions (e.g. allowing subzones to be managed by non-admin users) from the `stg` and `prd` zones. This is needed under the current AWS model, but not under the current GCP model.

Another advantage of a dedicated `dev` zone is that it allows use of a wildcard SSL certificate (other than `*.gpii.net`). This might have been nice for using Amazon Certificate Manager with dev clusters. It may still be nice if/when we get to TLS encryption for internal network communications.

However, I don't know if these factors warrant the extra handling needed to distinguish between "tyler.dev.gpii.net" and "prd.gpii.net" -- it's certainly a question worth asking. What does the team think?

mrtyler commented:
Note for @amatas (and @mrtyler ;)): I have replied to all the comments that I had replies for. I'm waiting to review code until things settle down a bit (otherwise it's hard for me to keep track of what has/hasn't been addressed). Please add a comment when you've taken at least a first pass at everything.

amatas commented:
OK team, I think that this is ready for a second round of review if you all believe that it's needed. @natarajaya @mrtyler @seg10 @stepanstipl , thanks for your thoughts and code suggestions, feel also free to review the code and add any comments.

I've applied and tested almost all the corrections suggested. I've tested parts of the code using my personal account, but it was not possible for me to test all together due to the dependency of cloud providers and accounts involved.

The split or the reorganization of the rake files I think that could be addressed in a separate ticket, as is not critical for this work and I wouldn't like to grow the size of this PR. The same to the READMEs organization.

This time I have used the merge instead of rebasing the code. I read that using this method the history of the changes makes sense as you can see the original code that a commit patches, meanwhile using the rebase method you lost that original code. Also it seems that Github manages better the comments in the code of the PR.

@stepanstipl @mrtyler I'd vote for using what we have now, developers domains hanging from a `dev` zone. I think that is an advantage the dedicated permissions of that non critical zone. But as always, we can see how it goes. I think that a swap to the other model is not so difficult.

@stepanstipl This PR doesn't address the deleting of the projects. This is a good feature for a new task (which I'll create as well as others commented in this PR), but as I said before I'd like to merge this code by now and stop adding features to this PR.

My plan for applying this code in the running environments is this:

1. Merge the gcp/rakefiles/entrypoint.rake as it is the only file that has conflicts with the actual master.
1. Merge the PR
1. Checkout the code in a live session with other team members.
1. Import the resources that are already created in GCP and match the patterns used in this code:
   1. Projects
   1. Service accounts
   1. User accounts permissions
   1. Buckets
   1. DNS zones
   1. DNS registries
1. Check the `plan` output and see that it doesn't break or delete current resources.
1. Run `rake infra_init`
1. Check the resources updated and created
1. Run `rake project_init`
1. Run `rake apply_projects`
1. Check the resources created, paying attention to the DNS zones structure and the NS registries.

amatas commented:
@mrtyler I added some commits to address your suggestions. Please check that everything is fine.

mrtyler commented:
LGTM!

Let's talk tomorrow about how to deploy this branch!

stepanstipl commented:
Rake will fail with following error when running `rake auth`:
```
rake aborted!
Errno::ENOENT: No such file or directory @ rb_sysopen - ../../.config/prd/compose.env
/Users/stepan/Projects/gpii/gpii-test/gcp/rakefiles/entrypoint.rake:48:in `initialize'
/Users/stepan/Projects/gpii/gpii-test/gcp/rakefiles/entrypoint.rake:48:in `open'
/Users/stepan/Projects/gpii/gpii-test/gcp/rakefiles/entrypoint.rake:48:in `block in <top (required)>'
Tasks: TOP => auth => set_vars
(See full trace by running task with --trace)
```

It's expecting `compose.env` that has not been created.

Steps to reproduce:
- Clone repo to a fresh new location
- Follow `common/README.md` steps "Creating the initial infrastructure"

stepanstipl commented:
Readme is reffereing to "../aws#configure-cloud-provider-credentials" for AWS creds setup, and that is mentioning 2 opitons - `aws configure` or manually putting creds in `~/.aws`. Exekube on the other hand seems to expect credentials in `./.config/${ENV}/aws`, which is used as docker-compose mount `/root/.aws`.

amatas commented:
thanks @stepanstipl I fixed both issues in https://github.com/gpii-ops/gpii-infra/pull/60/commits/f7c2d222ecdf13acf37c4e304d38f02bc74a4aff and https://github.com/gpii-ops/gpii-infra/pull/60/commits/514e66525c0d271cb46d233d369c3c7b8148fe43

stepanstipl commented:
@amatas is there a way to test the AWS DNS part? The zone `Z26C1YEN96KOGI` seems to be hardcoded.

stepanstipl commented:
`project_init` rake task fails with insufficient permissions:

```
module.gcp_zone.google_dns_managed_zone.main: Creating...
  description:    "" => "gcp DNS zone"
  dns_name:       "" => "gcp.stgca02.stipl.net"
  name:           "" => "gcp-gpii-net"
  name_servers.#: "" => "<computed>"
  project:        "" => "<computed>"
Releasing state lock. This may take a few moments...

Error: Error applying plan:

1 error(s) occurred:

* module.gcp_zone.google_dns_managed_zone.main: 1 error(s) occurred:

* google_dns_managed_zone.main: Error creating DNS ManagedZone: googleapi: Error 403: Forbidden, forbidden

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```
I assume it's because the `projectowner@gpii2-common-prd.iam.gserviceaccount.com` service account has only `Storage Admin` and `Viewer` permissions and therefore is unable to create the DNS zone.

stepanstipl commented:
Terraform is checking and downloading provides every time it's being run:
```
Initializing provider plugins...
- Checking for available provider plugins on https://releases.hashicorp.com...
- Downloading plugin for provider "google" (1.16.2)...
```

This is slow and can mean loss of consistent environemnt - if new version is found, it will be downloaded.

We shoudl
a) Specify versions of providers to lock these down
b) Predownload providers when we build the docker image and prevent any further downloads

stepanstipl commented:
While running `rake apply_projects` I get following credentials error:
```
[terragrunt] [/project/live/prd/infra/prd/zone] 2018/08/14 16:34:39 Running command: terraform apply -auto-approve -input=false -auto-approve
Acquiring state lock. This may take a few moments...
Releasing state lock. This may take a few moments...

Error: Error running plan: 1 error(s) occurred:

* provider.google: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.


[terragrunt] [/project/live/prd/infra/prd/zone] 2018/08/14 16:34:42 Module /project/live/prd/infra/prd/zone has finished with an error: Hit multiple errors:
exit status 1
[terragrunt] 2018/08/14 16:34:42 Encountered the following errors:
Hit multiple errors:
exit status 1
FATA[0083] exit status 1
rake aborted!
```
... atm. looking into why is this happening

amatas commented:
Hi @stepanstipl ,

1. DNS Route53 zone harcoded:
I changed the hardcoded value by a variable, so you can now set that value using TF_VAR environment variables and [the rake code should inject it in to the container](https://github.com/gpii-ops/gpii-infra/blob/fd32dc5c0f851138159ff4f2ae8850f6397d3e81/gcp/rakefiles/entrypoint.rake#L44-L51) and pass it to the terraform code.

1. DNS privileges in common project
I added the code to set the admin role to the SA

1. Terraform providers downloaded on every execution.
I don't know why this happens, it should save the files in the `.config` path. Anyway, your idea to save execution time could be a good improvement, I would file a jira for that.

1. Default credentials
I also saw that problem in my tests, I solved running `rake auth` again. And I wasn't able to figure out what happened before that raised that error. Perhaps a lack of correct permissions on the `credentials.db` file?

Thanks again for your comments and testing.

stepanstipl commented:
@amatas Cool, thanks Alfredo

I think the problem might be with using just `gcloud auth`, I think correct one should be `gcloud auth application-default login`. See https://www.terraform.io/docs/providers/google/index.html. Unfortunately in my case the problem is not fixed by running `rake auth` again.

stepanstipl commented:
I think I've figured it out.

The first module `/project/live/prd/infra/dns-root` which is pointing to `/project/modules//aws-gcp-dns` and works fine has following provider declaration:
```
:
```

The second, failing one, `/project/live/prd/infra/prd/zone` which is pointing to `/project/modules//gcp-project` does not have any such provider declaration.

The first one, therefore, uses explicitly defined credentials, while the second one is trying to use default ones as described at https://www.terraform.io/docs/providers/google/index.html.

amatas commented:
Good catch @stepanstipl ! I've added such provider to gcp-project module on https://github.com/gpii-ops/gpii-infra/pull/60/commits/c09bd448064dbd92e010d46c300de270c4cdbdde

stepanstipl commented:
I have hit another issue, when the service account for actual projects `serviceAccount:projectowner@gpii-common-prd.iam.gserviceaccount.com` in `modules/gcp-project/main.tf` is hard-coded and does not respect the actual "common" project name.  I would also prefer to have the other members parametrised (`group:ops@raisingthefloor.org`), instead of hardcoded in the module.

stepanstipl commented:
Atm. running `rake apply_project` always ends up in following error: 
```
[terragrunt] [/project/live/prd/infra/prd/zone] 2018/08/15 15:50:08 Running command: terraform apply -auto-approve -input=false -auto-approve
google_project.project: Refreshing state... (ID: gpii-gcp-prd2)

Error: Error refreshing state: 1 error(s) occurred:

* google_project.project: 1 error(s) occurred:

* google_project.project: google_project.project: Error retrieving App Engine application "gpii-gcp-prd2": Get https://appengine.googleapis.com/v1/apps/gpii-gcp-prd2?alt=json: dial tcp [2a00:1450:4009:80f::200a]:443$
 connect: cannot assign requested address


[terragrunt] [/project/live/prd/infra/prd/zone] 2018/08/15 15:50:15 Module /project/live/prd/infra/prd/zone has finished with an error: Hit multiple errors:
exit status 1
[terragrunt] 2018/08/15 15:50:15 Encountered the following errors:
Hit multiple errors:
exit status 1
FATA[0021] exit status 1
rake aborted!
```
... investigating why

mrtyler commented on common/modules/aws-gcp-dns/aws-dns-zone/main.tf:
```diff
@@ -10,15 +12,15 @@ resource "aws_route53_zone" "main" {
 
 resource "aws_route53_record" "main_ns" {
-  zone_id = "Z26C1YEN96KOGI"  # Unmanaged route53 zone for gpii.net
+  zone_id = "${var.aws_zone_id}"  # Unmanaged route53 zone for gpii.net
```
The comment is about the value, not the code, so can you move the comment to where you moved the value?

mrtyler commented on common/modules/gcp-project/main.tf:
```diff
@@ -33,9 +46,11 @@ resource "google_project" "project" {
     "compute.googleapis.com",
     "logging.googleapis.com",
     "monitoring.googleapis.com",
+    "appengine.googleapis.com",
```
This seems like a pretty weird API to need, although I see Stepan is having a problem related to appengine in [this comment](https://github.com/gpii-ops/gpii-infra/pull/60#issuecomment-413242122).

Do we have any idea what is pulling in a dependency on GAE?

stepanstipl replied:
I ended up with quite a long list of APIs, which in my experience were required once you want GKE cluster, but `appengine.googleapis.com` wasn't one of those. 

I'm wondering if this will eventually start fighting with each other once the GKE cluster is created - code here removing APIs, code for GKE enabling them again... - but I didn't get that far yet.

Can't figure out what is causing my error - if I try to import that very same GCP project as TF resource somewhere else, everything works fine, but whenever I run our code I end up with the very same error. I guess I'll try to disable the appengine API - might be a good hint :).

Just fyi the APIs I needed for working GKE were:
```
    "compute.googleapis.com",
    "container.googleapis.com",
    "cloudbilling.googleapis.com",
    "cloudresourcemanager.googleapis.com",
    "dns.googleapis.com",
    "iam.googleapis.com",
    "serviceusage.googleapis.com",
    "storage-api.googleapis.com",
    "oslogin.googleapis.com",
    "bigquery-json.googleapis.com",
    "containerregistry.googleapis.com",
    "pubsub.googleapis.com",
    "deploymentmanager.googleapis.com",
    "replicapool.googleapis.com",
    "replicapoolupdater.googleapis.com",
    "resourceviews.googleapis.com",
```

mrtyler commented on gcp/rakefiles/entrypoint.rake:
```diff
@@ -32,17 +32,20 @@ task :set_vars do
   ENV.each do |key, val|
    tf_vars << key if key.match(/^TF_VAR_/)
   end
+  Vars.create_dir_if_not_exists(@dot_config_path)
```
Can you use `directory` (plus a task-level dependency on `@dot_config_path`) instead?

stepanstipl commented:
Creating DNS zones fails with project named `prd2`. It's trying to create zone `prd.gcp.gpii.net.2` and that is not valid zone name. 

I guess the problem comes from the `/([a-z]+)-([a-z]+)-([a-z]+)-?([a-z]+)?/` regexp, which does not allow numbers in individual parts of the name.

```
google_dns_managed_zone.project: Creating...
  description:    "" => "gpii-gcp-prd2 DNS zone"
  dns_name:       "" => "prd.gcp.gpii.net.2"
  name:           "" => "gpii-gcp-prd2-zone"
  name_servers.#: "" => "<computed>"
  project:        "" => "gpii-gcp-prd2"
```

amatas commented:
Hi @stepanstipl 

1. the projectowner should have owner permissions in each project: https://github.com/gpii-ops/gpii-infra/blob/65a8f02d1e2f6dc99c704c45271a988f4944a841/common/modules/gcp-project/main.tf#L64-L73
and also it has permissions on the common-prd project https://github.com/gpii-ops/gpii-infra/blob/65a8f02d1e2f6dc99c704c45271a988f4944a841/common/rakefiles/entrypoint-common.rake#L46-L48
I don't understand where it needs more privileges.
1. Changed the regular expression to allow digits in https://github.com/gpii-ops/gpii-infra/pull/60/commits/65a8f02d1e2f6dc99c704c45271a988f4944a841

stepanstipl commented:
@amatas thanks, but looks like the last commit introduced an error:
```
[terragrunt] [/project/live/prd/infra/prd/zone] 2018/08/16 15:08:30 Running command: terraform apply -auto-approve -input=false -auto-approve

Error: Error parsing /root/.terragrunt/QwL5tE0zt_C6_1_1Ybv6FwtLII4/exDWDCWoRRnBaZ-HQfLUjTVcEno/gcp-project/main.tf: At 32:47: illegal char escape
```

Re. the storage permission - I think it's probably related to the previous one, when the service account name does not respect the project name - even if the main project is named `gpii2-common-prd`, as you can't have 2 projects with same names on GCP, the service account is still hadrcoded to  `projectowner@gpii-common-prd.iam.gserviceaccount.com`. So ignore that, it'll get fixed once the svc account name matches the project name.

stepanstipl commented:
cool @amatas, looks good. I started creating GCP org we can use for testing and will look into creating a GitLab task, I think this we'll also need to coordinate with work Tyler is doing with docker volumes #92

stepanstipl commented:
@amatas I'm trying to run stg, under the new shared test org, with some minor code changes - see https://github.com/stepanstipl/gpii-infra/tree/gpii-3125-stg - and following env variables set:
```
TF_VAR_aws_zone_id=Z1OFSNK98Q5PP7
TF_VAR_organization_domain=test1.gpii.net
TF_VAR_organization_id=327626828918
TF_VAR_organization_name=gpii2test
ORGANIZATION_ID=327626828918
```
(aws zone is mine test zone, didn't want to mess up our aws)

and I'm hitting following error, when rake is trying to create service account key, without creating the service account first - that part is getting skipped:
```
rake infra_init
(in /Users/stepan/Projects/gpii/gpii-infra/common/live/stg)

docker-compose run --rm --service-ports xk sh -c 'gcloud iam service-accounts keys create $TF_VAR_serviceaccount_key       --iam-account projectowner@$T$
_VAR_project_id.iam.gserviceaccount.com'
ERROR: (gcloud.iam.service-accounts.keys.create) PERMISSION_DENIED: Permission iam.serviceAccountKeys.create is required to perform this operation on service
 account projects/-/serviceAccounts/projectowner@gpii2test-common-stg.iam.gserviceaccount.com.
FATA[0001] exit status 1
rake aborted!
Command failed with status (1): [
    docker-compose run --rm --service-por...]
/Users/stepan/Projects/gpii/gpii-infra/gcp/rakefiles/entrypoint.rake:108:in `block in <top (required)>'
/Users/stepan/Projects/gpii/gpii-infra/common/rakefiles/entrypoint-common.rake:25:in `block in <top (required)>'
Tasks: TOP => set_current_project => secrets/kube-system/owner.json
(See full trace by running task with --trace)
```

The permission error is misleading, problem is that the service account does not exist (and does not get created), the user has permission to create both service account and a new key.

The account is not there:
```
gcloud iam service-accounts list --project gpii2test-common-stg
Listed 0 items.`
```

stepanstipl commented:
I have one issue I'm hitting when running `rake apply_projects`:
```
* google_project_iam_binding.project: Error applying IAM policy for project "gpii2test-gcp-stg": Error setting IAM policy for project "gpii2test-gcp-stg": go
ogleapi: Error 400: Request contains an invalid argument., badRequest
```

The issue is when trying to set `ops@raisingthefloor.org` as `roles/owner` in IAM policy. Owner role can only be taken by service account or user (see https://cloud.google.com/resource-manager/reference/rest/v1/projects/setIamPolicy).

A solution is to remove `group:ops@raisingthefloor.org` from `modules/gcp-project/main.tf`, this kind of permission should be inherited from org anyway, so not needed.

stepanstipl commented:
Another error I hit is `Error 403: Permission denied` when creating the NS record for the project. The error output is following:
```
* google_dns_record_set.ns-root: 1 error(s) occurred:

* google_dns_record_set.ns-root: Error retrieving record sets for "gcp-test1-gpii-net": googleapi: Error 403: Permission denied on resource project gpii2test-gcp-common-stg., forbidden

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.


[terragrunt] [/project/live/stg/infra/stg/zone] 2018/08/23 16:20:58 Module /project/live/stg/infra/stg/zone has finished with an error: Hit multiple errors:
exit status 1
[terragrunt] 2018/08/23 16:20:58 Encountered the following errors:
Hit multiple errors:
exit status 1
FATA[0021] exit status 1
rake aborted!
Command failed with status (1): [docker-compose run --rm --service-ports xk...]
```

Strangely when impersonating the `projectowner@gpii2test-common-stg.iam.gserviceaccount.com` service account, via `gcloud auth activate-service-account` and creating the DNS record manually, everything works as expected.

mrtyler commented on shared/rakefiles/vars.rb:
```diff
@@ -81,7 +85,7 @@ def self.set_versions()
     end
   end
-  def self.create_dir_if_not_exists(path)
+  def self.create_directory_if_not_exists(path)
```
Doesn't look like anything uses this method anymore. Can you remove it and its associated tests?

mrtyler commented on common/README.md:
```diff
@@ -4,54 +4,43 @@ This directory manages GPII infrastructure in [Google Cloud Project (GCP)](https
+
+The project structure is like following:
+
+- gpii-common-prd (only for creating the rest of the infrastructure)
```
Add a note about gpii-common-stg?

amatas replied:
Done on https://github.com/gpii-ops/gpii-infra/pull/60/commits/35b4ba9f9657b0c25e23c0f0d488d386453c7dd6

mrtyler commented on common/modules/gcp-project/main.tf:
```diff
@@ -0,0 +1,119 @@
+              replace(
+                google_project.project.name,
+                "/([\\w]+)-([\\w]+)-([\\w]+)-?([\\w]+)?/",
+                "$4.$3.$2.${var.organization_domain}."),
```
I would expect `google_project.project.name` to already have the organization domain (`gpii` or `testDomain` or whatever). Hence, replacing that value with `$(var.organization_domain}` seems incorrect vs using `$1`. Am I missing something here?

amatas replied:
The name of the testing organization is "gpii2test" and the domain used for that organization is "test1.gpii.net". A sample of the name of the projects created is `gpii2test-gcp-dev-joe`, which doesn't match with the domain name.

In the case of the GPII projects, the project name includes the `gpii` string of the `gpii.net`, we assumed the `.net`. That's why I introduce the `organization_domain` in the code, because the name of the organization may not match the domain name, and we use the name of the organization to give a name for the projects.

mrtyler commented on common/modules/gcp-project/main.tf:
```diff
@@ -0,0 +1,119 @@
+  ttl          = 3600
+  project      = "${var.organization_name}-gcp-${element(split("-", var.project_name), 0)}"
+  rrdatas      = ["${google_dns_managed_zone.project.name_servers}"]
+  count        = "${length(split("-", var.project_name)) == 2 ? 1 : 0}"
```
Will this work with `dev-username-with-lots-of-dashes`?

mrtyler commented on shared/rakefiles/entrypoint-common.rake:
```diff
@@ -0,0 +1,23 @@
+  # Due to the needed permissions for creating the following resources, only an
+  # administrator of the organization can run this task.
+
+  if @project_type != "common"
```
Can you remove this check?

mrtyler commented on shared/rakefiles/entrypoint-common.rake:
```diff
@@ -0,0 +1,23 @@
+    " run ",
+    " run --volume #{ENV["HOME"]}/.aws:/aws-backup ",
+  )
+  sh "#{@exekube_cmd_with_volume} sh -c '\
```
Can you copy the existing pattern for this in ci.rake?

mrtyler commented on shared/rakefiles/entrypoint.rake:
```diff
@@ -70,21 +72,15 @@ task :configure_serviceaccount => [:set_vars] do
-  sh "#{@exekube_cmd} rake configure_kubectl"
-end
+desc "Get the credentials needed to create all the resources inside the project"
+task :project_init => [:set_vars, :configure_current_project, :configure_serviceaccount]
```
Is it expected that this task has no contents?

amatas replied:
Yes, it is expected. This task "executes" the list of dependencies.

mrtyler commented on shared/rakefiles/tests/spec/vars_spec.rb:
```diff
@@ -109,6 +143,16 @@ def scrub_env
     expect(ENV).to have_received(:[]=).with("TF_VAR_nonce", a_value)
   end
+
+  it "Create directory path" do
```
Delete.

amatas replied:
sorry @mrtyler I can not understand this.

amatas replied:
Deleted, I used this function in the implementation of a solution that didn't work as expected.

mrtyler commented on shared/rakefiles/xk.rake:
```diff
@@ -0,0 +1,131 @@
+      --organization #{ENV["ORGANIZATION_ID"]} \
+      --set-as-default"
+  else
+    sh "#{@exekube_cmd} gcloud config set project #{ENV["TF_VAR_project_id"]}"
```
Can you use `:configure_current_project` for this?

amatas replied:
the file `xk.rake` is the only rake file used inside the Docker container, the task `:configure_current_project` is defined in the `entrypoint.rake`. Do you think that is worth to merge these two files?. If so, perhaps it could be addressed in another Jira task.

mrtyler commented on shared/rakefiles/xk.rake:
```diff
@@ -0,0 +1,131 @@
+
+  Rake::Task[:configure_serviceaccount].invoke
+
+  sh "#{@exekube_cmd} gcloud projects add-iam-policy-binding #{ENV["TF_VAR_project_id"]} \
```
How about:

```
["roles/viewer", "roles/storage.admin", "roles/dns.admin"].each do |role|
  sh "... --role #{role}"
end
```

mrtyler commented on shared/rakefiles/xk.rake:
```diff
@@ -0,0 +1,131 @@
+     sh "#{@exekube_cmd} gcloud services enable #{service}" unless hash.any? { |s| s['serviceName'] == service }
+  end
+
+  sh "#{@exekube_cmd} gcloud organizations add-iam-policy-binding #{ENV["ORGANIZATION_ID"]} \
```
I would do a `.each` for these two commands as well.

stepanstipl commented on common/README.md:
```diff
@@ -4,54 +4,43 @@ This directory manages GPII infrastructure in [Google Cloud Project (GCP)](https
+1. By default you'll use the RtF Organization and Billing Account.
+   * You can use a different Organization or Billing Account, e.g. from a GCP Free Trial Account, with `export ORGANIZATION_ID=111111111111` and/or `export BILLING_ID=222222-222222-222222`.
+1. Check that [you have the AWS credentials](../aws#configure-cloud-provider-credentials).
+   * Copy the `credentials` file to the `common/.config/prd/aws` directory.
```
I think this is not correct anymore, looks like the new volume code always expect the credentials to be at `~/.aws` - this seems inconsistent with the rest?

stepanstipl commented:
I still do get the "Error retrieving App Engine application" error mentioned above.

Also I see some flapping of services being enabled/disabled on subsequent runs:
```
google_project_services.project: Modifying... (ID: gpii2test-gcp-stg)
  services.#:          "8" => "6"
  services.1568433289: "oslogin.googleapis.com" => "oslogin.googleapis.com"
  services.1633256984: "appengine.googleapis.com" => "appengine.googleapis.com"
  services.1954675454: "serviceusage.googleapis.com" => ""
  services.2240314979: "compute.googleapis.com" => "compute.googleapis.com"
  services.2928564140: "dns.googleapis.com" => "dns.googleapis.com"
  services.3237295688: "monitoring.googleapis.com" => "monitoring.googleapis.com"
  services.3355193353: "logging.googleapis.com" => "logging.googleapis.com"
  services.3875785048: "storage-api.googleapis.com" => ""
```

Also this will fight with services required for Kubernetes cluster, as when actual GKE cluster is created, it will enable several additional services and when this code is run, it'll try to disable those again.

Previously I ended up requiring all these services to be enabled on projects for GKE to work:
```
  services = [
    "compute.googleapis.com",
    "container.googleapis.com",
    "cloudbilling.googleapis.com",
    "cloudresourcemanager.googleapis.com",
    "dns.googleapis.com",
    "iam.googleapis.com",
    "serviceusage.googleapis.com",
    "storage-api.googleapis.com",
    "oslogin.googleapis.com",
    "bigquery-json.googleapis.com",
    "containerregistry.googleapis.com",
    "pubsub.googleapis.com",
    "deploymentmanager.googleapis.com",
    "replicapool.googleapis.com",
    "replicapoolupdater.googleapis.com",
    "resourceviews.googleapis.com",
  ]
```

stepanstipl commented:
👍 Looks good, I was able to run it without any errors this time. Yeey!

Two questions/concerns:
- How should we behave when deleting end? Currently, after I created a new project (folder in `common/live/stg/infra/dev/`), ran `rake apply_infra`, and deleted the project afterwards, the automation just stops caring about it and when you run `rake apply_infra` again all the resources are left to hang there.

- Is this expected to work with current gcp code that actually creates the cluster? (I Haven't that yet)

stepanstipl commented:
I keep hitting (even after retrying) following error with the latest version of the code:

```
Error: Error refreshing state: 1 error(s) occurred:

* google_dns_managed_zone.project: 1 error(s) occurred:

* google_dns_managed_zone.project: google_dns_managed_zone.project: Error reading DNS Managed Zone "dev-gcp-test1-gpii-net": googleapi: Error 403: Google Cloud DNS API has not been used in project 170800538300 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/dns.googleapis.com/overview?project=170800538300 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry., accessNotConfigured


[terragrunt] [/project/live/stg/infra/dev/billy] 2018/09/03 11:54:35 Module /project/live/stg/infra/dev/billy has finished with an error: Hit multiple errors:
exit status 1
```

and it looks like that at this stage the DNS api is indeed not enabled on the porject:
```
$ gcloud --project gpii2test-gcp-dev-joe services list
NAME                          TITLE
bigquery-json.googleapis.com
cloudbilling.googleapis.com
compute.googleapis.com
oslogin.googleapis.com
pubsub.googleapis.com
serviceusage.googleapis.com
storage-api.googleapis.com
```

stepanstipl commented:
Oh sorry if this wasn't clear, this happens when running the `rake apply_infra` of the common part.

stepanstipl commented:
Looks good (after enabling the DNS api), but now the latest version of code fails on:

```
$ rake apply_infra                                                                                                                                                                                     
rake aborted!
LoadError: cannot load such file -- /Users/stepan/Projects/gpii/gpii-infra/shared/rakefiles/test.rake

(See full trace by running task with --trace)
```

amatas commented:
@stepanstipl I left behind the test.rake file in the last merge

mrtyler commented:
Github won't let me reply to [this thread](https://github.com/gpii-ops/gpii-infra/pull/60#discussion_r214666740).

In https://github.com/gpii-ops/gpii-infra/pull/60#discussion_r213119778 I suggested deleting `Vars. create_directory_if_not_exists` since nothing calls it.

My comment in https://github.com/gpii-ops/gpii-infra/pull/60#discussion_r214666740 was suggesting to also delete the test for `Vars. create_directory_if_not_exists`.

I hope this clears things up.

mrtyler commented:
Github won't let me reply to [this thread](https://github.com/gpii-ops/gpii-infra/pull/60#discussion_r214666689).

> the file xk.rake is the only rake file used inside the Docker container, the task :configure_current_project is defined in the entrypoint.rake. Do you think that is worth to merge these two files?.

Not merged, but you could follow the [existing pattern](https://github.com/gpii-ops/gpii-infra/blob/master/gcp/rakefiles/entrypoint.rake#L75).

mrtyler commented on shared/rakefiles/xk.rake:
```diff
@@ -65,6 +65,37 @@ task :xk, [:cmd, :skip_infra, :skip_secret_mgmt] => [@serviceaccount_key_file, @
+  # conflicts at the creation resources we need to import the already created
+  # DNS zone.
+
+  dns_zone = @env + "-" + args[:project_type] + "-" + ENV["TF_VAR_organization_domain"].tr(".", "-")
```
Since you're doing this calculation in Ruby code as well as in several places in Terraform code, how about doing it once in `set_vars()` and referring to the result of that calculation where needed?

mrtyler commented on shared/rakefiles/xk.rake:
```diff
@@ -65,6 +65,37 @@ task :xk, [:cmd, :skip_infra, :skip_secret_mgmt] => [@serviceaccount_key_file, @
   sh_filter "#{@exekube_cmd} #{args[:cmd]}" if args[:cmd]
 end
 
+task :refresh_infra, [:project_type] => [@gcp_creds_file] do | taskname, args|
```
I find this name confusing since each gcp/live environment has its own `infra` directory. How about `refresh_common_infra`?

mrtyler commented on gcp/modules/gpii-preferences/main.tf:
```diff
@@ -18,6 +23,8 @@ data "template_file" "preferences_values" {
 
   vars {
     env                    = "${var.env}"
+    # TODO: remove one of the following variables
```
What does this mean? Can you remove one now?

mrtyler commented on common/README.md:
```diff
@@ -37,10 +37,16 @@ The DNS zones are:
 1. `rake apply_infra`
    * This will create all the projects in the organization. Each project is defined by the content of a directory in `common/live/(stg|prd)/infra`
+
+WARNING: The command `rake destroy_infra` of the GCP part of this project can disable the DNS API driving to an issue where Terraform is unable to refresh the state at the next executions. Avoid the use of `rake destroy_infra`, and only remove the most expensive resources using `rake destroy`.
```
This seems like a pain. Assuming we can't fix it now, can you create a ticket and reference that ticket from here?

amatas replied:
Ticket created: https://issues.gpii.net/browse/GPII-3332

mrtyler commented on shared/rakefiles/xk.rake:
```diff
@@ -65,6 +65,37 @@ task :xk, [:cmd, :skip_infra, :skip_secret_mgmt] => [@serviceaccount_key_file, @
+  end
+end
+
 task :infra_init => [@gcp_creds_file] do
```
I find this name confusing since each gcp/live environment has its own infra directory. How about `apply_common_infra`?

stepanstipl commented:
Latest version works fine for me. I still have the question re. deleting projects - any thoughts on how this should work? (check one of my previous comments for details)

amatas commented on .gitlab-ci.yml:
```diff
@@ -119,6 +137,30 @@ gcp-promote-to-stg:
+common-stg:
+  stage: common-stg
+  tags:
+    - common
```
Yes, I did

amatas commented on .gitlab-ci.yml:
```diff
@@ -119,6 +137,30 @@ gcp-promote-to-stg:
+  tags:
+    - common
+  variables:
+    ORGANIZATION_ID: "327626828918"
```
It is needed by the second part of the job, where a dev environment (dev-doe) is created in the testing organization using the `gcp` code.

amatas replied:
And I only wanted to be sure that the common-stg is always applied in the testing organization.

mrtyler commented on .gitlab-ci.yml:
```diff
@@ -153,7 +153,7 @@ common-stg:
     - rake infra_init
     - rake apply_infra
-    - cd ../../../gcp/live/stg
+    - cd ../../../gcp/live/dev
```
I don't understand this change.

I also think it's pretty unexpected to deploy an entire gcp/dev environment in the middle of a common/stg deployment. Why are we doing this? Is there a different approach we can take?

amatas replied:
The goal of the `stg` environment of the common part is the testing of the initialization of the organization, and whether all the resources created are able to support the dev/stg/prd environments. I think that a dev environment is the worst case scenario where multiple subdomains (also set by the initial resources) are involved. That's why a dev environment of a virtual user is created. We can also add the creation of a stg or prd environment if needed.
And all of the above avoiding any interaction with the real production organization (*.gcp.gpii.net).

stepanstipl commented:
I've noticed a lot of errors like the one below on the cluster created in the project created by this code:

```
E0906 14:33:27.288832       1 stackdriver.go:58] Error while sending request to Stackdriver googleapi: Error 403: Stackdriver Monitoring API has not been used in project 170800538300 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/monitoring.googleapis.com/overview?project=170800538300 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry., accessNotConfigured
```

Looks like we're missing Stackdriver API. The errors are from `prometheus-to-sd-exporter`.

amatas commented on shared/rakefiles/xk.rake:
```diff
@@ -65,6 +65,37 @@ task :xk, [:cmd, :skip_infra, :skip_secret_mgmt] => [@serviceaccount_key_file, @
+    terragrunt state show module.gke_network.google_dns_managed_zone.dns_zones \
+    --terragrunt-working-dir /project/live/#{@env}/infra/network \
+    "`
+  id_found = output[/^id *= *([\w-]*$)/,1]
```
Done, I've used the output command that supports the json output.

stepanstipl commented:
With the latest code I'm getting the following error:
```
module.gke_network.google_dns_managed_zone.dns_zones: Importing from ID "joe-dev-gcp-test1-gpii-net"...
module.gke_network.google_dns_managed_zone.dns_zones: Import complete!
  Imported google_dns_managed_zone (ID: joe-dev-gcp-test1-gpii-net)
module.gke_network.google_dns_managed_zone.dns_zones: Refreshing state... (ID: joe-dev-gcp-test1-gpii-net)

Error: module.gke_network.google_dns_managed_zone.dns_zones (import id: joe-dev-gcp-test1-gpii-net): 1 error(s) occurred:

* import module.gke_network.google_dns_managed_zone.dns_zones result: joe-dev-gcp-test1-gpii-net: import module.gke_network.google_dns_managed_zone.dns_zones (id: joe-dev-gcp-test1-gpii-net): Terraform detected a resource with this ID doesn't
exist. Please verify the ID is correct. You cannot import non-existent
resources using Terraform import.


[terragrunt] 2018/09/07 15:36:11 Hit multiple errors:
exit status 1
FATA[0003] exit status 1
```

I believe it's because the name of the zone was changed:
`name         = "${local.dnsname}"` -> `name         = "${local.dnsname}."`

but this is not reflected in the import command.

stepanstipl commented:
Again first 2 times I run `apply_infra` I got the old error:
```
[terragrunt] [/project/live/stg/infra/dev/billy] 2018/09/10 15:24:56 Module /project/live/stg/infra/dev/billy has finished with an error: Cannot process module Module /project$
live/stg/infra/dev/billy (dependencies: [/project/live/stg/infra/dev/zone]) because one of its dependencies, Module /project/live/stg/infra/dev/zone (dependencies: [/project/li
ve/stg/infra/dns-root]), finished with an error: Hit multiple errors:
exit status 1

Error: Error refreshing state: 1 error(s) occurred:

* google_project_services.project: 1 error(s) occurred:

* google_project_services.project: google_project_services.project: failed to list services: Get https://serviceusage.googleapis.com/v1beta1/projects/gpii2test-gcp-prd/services
?alt=json&fields=services%2Fname%2CnextPageToken&filter=state%3AENABLED: dial tcp [2a00:1450:4009:801::200a]:443: connect: cannot assign requested address
```

and after that it works fine :) and I'm unable to reproduce the error in any way. Otherwise looks good.

mrtyler commented on .gitlab-ci.yml:
```diff
@@ -3,10 +3,28 @@ stages:
   - unit-tests
   - dev
   - promote-to-stg
```
(Github won't let me add this comment where I want it.)

Should we have `common-promote-to-stg` and `common-promote-to-prd` Jobs?

mrtyler commented on .gitlab-ci.yml:
```diff
@@ -3,10 +3,28 @@ stages:
   - dev
   - promote-to-stg
+  - common-stg
   - stg
```
(Github won't let me add this comment where I want it.)

At some point (maybe right now :)) it sounds like you'll want to enable the (currently disabled) `gcp-stg` and `gcp-prd` Jobs.

mrtyler commented:
Since we agreed to try to merge and deploy this PR tomorrow (Tuesday) at 1000 MDT, I propose:

1. @amatas, please address as many of my review comments as you can by 0800 MDT tomorrow.
2. I'll get online by 0800 MDT to review these changes.
3. Once we get those out of the way, we can discuss anything remaining from the "more significant changes" pile.
4. Based on that discussion, we can make a go/no-go decision ("Alfredo already made all the needed changes and/or we can make them together real quick and/or we decided they weren't needed" vs "Alfredo needs more time to implement the new CI strategy we worked out") by 0900 MDT when we have a meeting unrelated to this deployment.
5a. If we reach a "go" decision, we start deploying after the 0900 meeting, hopefully around 1000 MDT.
5b. If we reach a "no-go" decision, re-schedule deployment for Wednesday or Thursday.

mrtyler commented on shared/rakefiles/xk.rake:
```diff
@@ -68,38 +68,42 @@ task :xk, [:cmd, :skip_secret_mgmt] => [@serviceaccount_key_file, @kubectl_creds
-    terragrunt state show module.gke_network.google_dns_managed_zone.dns_zones \
-    --terragrunt-working-dir /project/live/#{@env}/infra/network \
+    terragrunt output -json dns_zones \
+    --terragrunt-working-dir /project/live/#{@env}/infra/network 2>/dev/null
```
Looks like you didn't do this one. Why not?

mrtyler commented on .gitlab-ci.yml:
```diff
@@ -119,6 +137,30 @@ gcp-promote-to-stg:
+    - rake clobber
+    - rake configure_serviceaccount_ci_restore
+    - rake
+    - rake destroy
```
`after_script` is maybe a step in the right direction. What will this do if there is a failure early in the `script` section and the dev environment is never created?

mrtyler commented:
LGTM!!

natarajaya commented on shared/rakefiles/ci.rake:
```diff
@@ -5,6 +5,10 @@ task :set_vars_ci => [:set_vars] do
     " run ",
```
@amatas Maybe I am missing something, but it is not clear to me why do we need another `shared` dir just for rake files? Our shared Helm charts live in `common/charts`, why don't have rake files in `common/rakefiles`?

amatas replied:
We need a separate directory outside the `common` and `gcp` directory where [these imports of the entrypoint.rake](https://github.com/gpii-ops/gpii-infra/blob/d2a4efb99a5dd9c02806d3ac16f92ff7f43244be/shared/rakefiles/entrypoint.rake#L3-L4) could work in both cases because it uses the same relative path. If the rake files live in the `common` directory the import would be different if you call the rakes from `common` and `gcp`, so in one of the cases will fail.

natarajaya replied:
@amatas If it is only about relative path for rake file imports, then this should do the trick:
```
@rakefiles_dir = File.dirname(__FILE__)

import "#{@rakefiles_dir}/ci.rake"
import "#{@rakefiles_dir}/test.rake"
```
I would be very happy if we get rid of additional `shared` dir and let our rake files live in `common`!

amatas replied:
@natarajaya could we move forward this PR and tackle this in a separate ticket?. This change was discussed at the beginning with @mrtyler (https://github.com/gpii-ops/gpii-infra/pull/60/commits/69a69b2cdb7d6f2f34c0ee829ff28a767c07a873) , and we found the solution of using a shared directory for shared resources like helm charts and rakefiles that are not tied directly with a particular environment (common is an environment itself with a particular pipeline). I wouldn't like to revert this work now.

Seriously, if the agreement is to have the rakefiles inside `common` I will be fine with that. But please, let's merge this PR first and revert this change if so later.

natarajaya commented on shared/rakefiles/entrypoint-common.rake:
```diff
@@ -0,0 +1,11 @@
+desc "[ONLY ADMIN] Initialize GCP provider where all the projects will live"
```
@amatas We use underline `_` as "space substitution" for other rake files: `ci_save_all.rake`, `sh_filter.rb`. I would suggest to use the same style.

natarajaya commented on shared/rakefiles/entrypoint.rake:
```diff
@@ -122,12 +111,11 @@ end
   else
     puts "Argument :cmd -- the command to run inside the exekube container -- not present, defaulting to bash"
-    cmd = "bash"
+    sh "#{@exekube_cmd} sh"
```
@amatas Why do we need these changes? `sh "#{@exekube_cmd} sh"` is not the same  as `sh "#{@exekube_cmd} rake xk['#{cmd}',skip_secret_mgmt]"`. With this, secrets will not be applied when you call `rake sh` without providing command, and shell output will not be obfuscated.

amatas replied:
I introduced this changes in order to avoid load functions if I only need a shell to debug inside the container. In which cases do we need to load the secrets when we invoke "rake sh"?

Before these changes, if I just wanted to test my TF code I had to authenticate against gcloud and do some stuff to get the SA, which in that time when I was debugging they didn't exist at GCP.

natarajaya replied:
@amatas Ok, I understand this now. Example of situation when we need secrets inside exekube shell is [CouchDB data restoration](https://github.com/gpii-ops/gpii-infra/tree/master/gcp#restoring-couchdb-data). Different behaviors for `rake sh` that simply invokes bash and for `rake sh[bash]` that does the same but via ruby code in container, is a bit confusing. Maybe introduce something like `:plain_sh` for that purpose? But I am okay leaving this as it is.

stepanstipl replied:
I would prefer to have secrets by default - you need these to be able to debug or work at all with TF state, without you can't do much (well anything with TF)

amatas replied:
I've leave the original task as `:rake_shell` and created a new one `:plain_shell` for covering those cases where someone wants to get into a shell without external dependencies

natarajaya commented on common/modules/gcp-project/main.tf:
```diff
@@ -0,0 +1,131 @@
+              replace(
+                google_project.project.name,
+                "/([\\w]+)-([\\w]+)-([\\w]+)-?([\\w]+)?/",
+                "$4.$3.$2.${var.organization_domain}"),
```
I am fine with any of those, but agree that if team decides to use "tyler-dev-gcp-gpii-net" naming convention, now is a good time for the change.

amatas replied:
We aggreed to decline this change.

natarajaya commented on shared/rakefiles/xk.rake:
```diff
@@ -0,0 +1,175 @@
+    dns_zone_found = false
+  end
+
+  if not dns_zone_found then
```
`unless` instead of `if not`, `then` is usually omitted in constructions like this in Ruby.

natarajaya commented on shared/rakefiles/xk.rake:
```diff
@@ -0,0 +1,175 @@
+  begin
+    # What does !! mean?
+    # https://stackoverflow.com/questions/524658/what-does-mean-in-ruby
+    dns_zone_found = !!JSON.parse(output)
```
Maybe it looks more proper, but you don't really need this:
```irb(main):001:0> require 'json'
=> true
irb(main):002:0> a = '{}'
=> "{}"
irb(main):003:0> a_found = JSON.parse(a)
=> {}
irb(main):004:0> puts 'Got ya!' if a_found
Got ya!```

amatas replied:
I know, but I like to use the proper variable type always I can.

natarajaya replied:
This is fine. But again, if the choice is to use the code that works and easy to understand, or the code that requires you to lookup stackoverflow in order to understand whats happening, I would choose the first option :)

natarajaya commented on shared/rakefiles/xk.rake:
```diff
@@ -0,0 +1,175 @@
+  sh_filter "#{@exekube_cmd} #{args[:cmd]}" if args[:cmd]
+end
+
+task :refresh_common_infra, [:project_type] => [@gcp_creds_file] do | taskname, args|
```
I don't think that `xk.rake` is a right place for `:refresh_common_infra` and `:apply_common_infra` tasks. It is intended to be a lightweight wrapper for commands that require secrets to be set prior to execution. Is there anything that prevents moving those 2 tasks to common entrypoint rake file?

amatas replied:
Both tasks don't require secrets, they only need to be authenticated in GCP, that's why the `@gcp_creds_file` is a dependency.

Also they are invoked from [entrypoint.rake](https://github.com/gpii-ops/gpii-infra/blob/d2a4efb99a5dd9c02806d3ac16f92ff7f43244be/shared/rakefiles/entrypoint.rake#L79) and [entrypoint_common.rake](https://github.com/gpii-ops/gpii-infra/blob/d2a4efb99a5dd9c02806d3ac16f92ff7f43244be/shared/rakefiles/entrypoint-common.rake#L10) by tasks that set the secrets.

But if secrets were needed, I'd vote for creating a new task inside xk.rake in order to manage them and then use it as a dependency by other tasks. I think that one goal is to have as much as possible inside the Docker container.

I can not see the advantages of adding an extra layer to the sh commands. I mean using `rake -> terraform -> sh` instead of `rake -> sh`.

natarajaya replied:
@amatas Ok, I am still convinced that those tasks belong to entrypoint and entrypoint_common. However, I understand if you don't want to deal with this now and focus on merging this PR instead. We can fix this later.

natarajaya commented on shared/rakefiles/xk.rake:
```diff
@@ -0,0 +1,175 @@
+      --role #{role}"
+  end
+
+  `#{@exekube_cmd} gsutil ls gs://#{ENV["TF_VAR_project_id"]}-tfstate`
```
Beside the fact that we don't use [backtick calls ](http://ruby-doc.org/core-1.9.3/Kernel.html#method-i-60) in other rake files, why don't just:
```
sh "#{@exekube_cmd} sh -c ' \
  if [ "$(gsutil ls -p #{ENV["TF_VAR_project_id"]} | grep tfstate)" == "" ]; then \
    gsutil mb gs://#{ENV["TF_VAR_project_id"]}-tfstate \
  fi \
  gsutil versioning set on gs://#{ENV["TF_VAR_project_id"]}-tfstate \
'"
```

amatas replied:
I don't mind to use your suggested code, but I can not see a real advantage: relay on sh conditionals or ruby, save some chars vs more backslashes. Let's the team vote: @stepanstipl @mrtyler @seg10 what do you think?

```
  `#{@exekube_cmd} gsutil ls gs://#{ENV["TF_VAR_project_id"]}-tfstate`
  if $?.exitstatus != 0
    sh "#{@exekube_cmd} gsutil mb gs://#{ENV["TF_VAR_project_id"]}-tfstate"
  end

  sh "#{@exekube_cmd} gsutil versioning set on gs://#{ENV["TF_VAR_project_id"]}-tfstate" 
```

or

```
sh "#{@exekube_cmd} sh -c ' \
  if [ \"$(gsutil ls -p #{ENV["TF_VAR_project_id"]} | grep tfstate)\" == \"\" ]; then \
    gsutil mb gs://#{ENV["TF_VAR_project_id"]}-tfstate \
  fi \
  gsutil versioning set on gs://#{ENV["TF_VAR_project_id"]}-tfstate \
'"
```

natarajaya replied:
Same as in [previous thread](https://github.com/gpii-ops/gpii-infra/pull/60#discussion_r217216485). Let's merge this PR to get things rolling, and work on improving this later.

natarajaya commented:
@amatas This looks very close to being merged! Please check my last bunch of comments!

stepanstipl commented:
Related to Sergey's comments (and more in Slack), I've hit an issue with `rake sh` behaving differently (not setting up secrets mgmt) in this branch, compared to the master.

amatas commented:
@stepanstipl @natarajaya about the `rake sh` and the secrets mgmt, I just wanted a way to get a shell inside the container without running too much stuff at the invocation. I'm open to any other alternative to get this.
What if you run `rake sh[bash]`? I think that the entry command will run all the rake stuff that you need as it is implemented on master.
